infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/msghandlers/LogReplicationServer.java

```diff
@@ -296,17 +298,27 @@ private void handleMetadataRequest(RequestMsg request, ResponseMsg res,
     }
 
     /**
-     * Given a leadership request message, send back a
-     * response indicating our current leadership status.
+     * Given a leadership request message, send back a response indicating our current leadership status.
+     * Since the client registration happens on the SOURCE, leadership query msg also serves as a vehicle to notify the
+     * SINK about the new sessions. Response to the leadership request msg is sent only after processing the session
+     * information in the msg header.
      *
      * @param request the leadership request message
      * @param router  router used for sending back the response
      */
     @LogReplicationRequestHandler(requestType = LR_LEADERSHIP_QUERY)
-    private void  handleLeadershipQuery(RequestMsg request, ResponseMsg res,
+    private void handleLeadershipQuery(RequestMsg request, ResponseMsg res,
                                                      @Nonnull IClientServerRouter router) {
         log.debug("Log Replication Query Leadership Request received by Server.");
         HeaderMsg responseHeader = getHeaderMsg(request.getHeader());
+        // if session in the leadership query is not known to the local cluster, create and process the new session information
+        if (replicationContext.getIsLeader().get()) {
+            LogReplicationSession session = getSession(request);
+            if (!sessionManager.getIncomingSessions().contains(session)) {
```

Comment:
we have to match the topology ids on the source and sink before proceeding.  The Sink can have processed a topology update where the remote source cluster is removed or changed.  In that case, the session should not be added.

Reply:
Functionally we don't have an issue because the data itself would not be written if the topology ID on SOURCE and SINK doesn't match.
Another way to look at it is that this check is deferred to lower layers at the time of writing the replicated data.

Adding the topology check would be a small optimization where we wouldn't create the session and SinkManager objects. But to achieve this, we'll need to include the topologyID in the leadership query msg and also add the logic to drop all the msgs for which there is no session at the SINK.

I think the gain is small but the logic added would be more. So I am inclined to keep it this way.

Please let me know if any reviewers are of different opinion